### PR TITLE
WIP: Base64 decoder rust/v1

### PR DESCRIPTION
**Pull request by inashivb**
Just a show of work. Failures are known and fixes are yet TBD.

Link to tickets:
- https://redmine.openinfosecfoundation.org/issues/7065
- https://redmine.openinfosecfoundation.org/issues/7058
- https://redmine.openinfosecfoundation.org/issues/6280

### CODE REVIEW

File: rust/src/utils/mod.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+pub mod scbase64;
```
Comment (inashivb) [inline rust/src/utils/mod.rs:1]: Q: Is this naming scheme ok?

File: rust/src/utils/scbase64.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        Base64Mode::RFC2045 => {
+            if let Ok(fin_str) = decode_rfc2045(&mut decoder, in_vec) {
+                fin_str
+            } else { Vec::new() }
+        },
+        Base64Mode::RFC4648 => {
+            if let Ok(fin_str) = decode_rfc4648(&mut decoder, in_vec, max_decoded) {
+                fin_str
+            } else { Vec::new() }
+        },
```
Comment (inashivb) [inline rust/src/utils/scbase64.rs:287]: Q: Is there a better way to do this?

File: src/datasets.c [hunk 3]
```diff
@@ -521,10 +521,16 @@ static int DatasetLoadString(Dataset *set)
...
-            if (code == BASE64_ECODE_ERR) {
+            uint32_t num_decoded = 0;
+            Base64Decoded *b64d = rs_base64_decode((const uint8_t *)line, strlen(line), 0, Strict);
+            if (b64d == NULL) {
+                FatalErrorOnInit("bad base64 encoding %s/%s", set->name, set->load);
+                continue;
+            }
+            memcpy(decoded, b64d->decoded, b64d->decoded_len);
+            num_decoded = b64d->decoded_len;
+            rs_base64_decode_free(b64d);
```
Comment (inashivb) [inline src/datasets.c:532]: Q: Is this an OK pattern?

File: src/util-base64.c [hunk 4]
```diff
@@ -274,19 +274,19 @@ static inline Base64Ecode DecodeBase64RFC4648(uint8_t *dest, uint32_t dest_size,
...
  */
 Base64Ecode DecodeBase64(uint8_t *dest, uint32_t dest_size, const uint8_t *src, uint32_t len,
-        uint32_t *consumed_bytes, uint32_t *decoded_bytes, DetectBase64Mode mode)
+        uint32_t *consumed_bytes, uint32_t *decoded_bytes, Base64Mode mode)
 {
     *decoded_bytes = 0;
     Base64Ecode ret = BASE64_ECODE_OK;
     switch (mode) {
-        case Base64ModeRFC4648:
+        case RFC4648:
```
Comment (inashivb) [inline src/util-base64.c:282]: I was trying to make use of the base64 namespace on the rust side but the code is also used in C.
Q: Is this naming scheme ok? Or, was previous better?

**Comment by suricata-qa** [comment]
ERROR:
 ERROR: panic in ASAN_TLPR1_suri QA test

ERROR: QA failed on ASAN_TLPR1_suri.
  

Pipeline 21450

**Pull request closed without merging**